### PR TITLE
CleanRoom improvements

### PR DESCRIPTION
**Pull request by MatthewARoy**
## Pull request summary
CleanRoom enhancements and suspended colony util function.
### Description:
This adds support to CleanRooms so that it will now automatically drill to get to a controller if it's blocked off. It will also collect resources from enemy structures if you have set it to.
Also adds a util function for suspended colonies.
### Added:
- Adds util to listSuspendedColonies()
- Add constant to decide how aggressively to collect resources.

### Changed:
CleanRoom will now dismantle to get to a controller.
CleanRoom will now grab resources left in storage/terminal of other players.

## Testing checklist:
Currently running in: https://screeps.com/a/#!/room/shard2/W19N55
https://screeps.com/a/#!/history/shard2/W19N55?t=14996000

### CODE REVIEW

File: src/console/Console.ts [hunk 1]
```diff
@@ -244,6 +245,17 @@ export class OvermindConsole {
 		}
 	}
 
+	static listSuspendedColonies(): string {
```
Comment (bencbartlett) [inline src/console/Console.ts:255]: nice

File: src/declarations/memory.d.ts [hunk 2]
```diff
@@ -1,4 +1,11 @@
 type operationMode = 'manual' | 'semiautomatic' | 'automatic';
+/**
+ * 0: Basic
+ * 1: Collect from enemy storage/terminal
+ * 2: Collect from all sources TBD
+ * 3: Collect all and mine walls for energy TBD
+ */
+type resourceCollectionMode = number;
```
Comment (bencbartlett) [inline src/declarations/memory.d.ts:8]: why do you think this needs its own setting?

File: src/directives/colony/clearRoom.ts [hunk 3]
```diff
@@ -29,6 +34,9 @@ export class DirectiveClearRoom extends Directive {
 						`removing directive!`);
 			this.remove(true);
 		}
+		if (Memory.settings.resourceCollectionMode && Memory.settings.resourceCollectionMode >= 1) {
```
Comment (bencbartlett) [inline src/directives/colony/clearRoom.ts:35]: use enums or constants for numeric settings

File: src/directives/colony/clearRoom.ts [hunk 4]
```diff
@@ -52,7 +60,9 @@ export class DirectiveClearRoom extends Directive {
 			for (let s of allStructures) {
 				if (s.structureType == STRUCTURE_CONTROLLER) continue;
 				if (keepStorageStructures &&
-					(s.structureType == STRUCTURE_STORAGE || s.structureType == STRUCTURE_TERMINAL)) {
+					(s.structureType == STRUCTURE_STORAGE || s.structureType == STRUCTURE_TERMINAL) && hasContents(s.store)) {
+					// Create a collection flag
+					let result = s.pos.createFlag(undefined, DirectiveHaul.color, DirectiveHaul.secondaryColor);
```
Comment (bencbartlett) [inline src/directives/colony/clearRoom.ts]: sigh... I'm gonna need to work out all the bugs in DirectiveHaul, aren't I 😆

File: src/directives/colony/clearRoom.ts [hunk 5]
```diff
@@ -83,6 +93,22 @@ export class DirectiveClearRoom extends Directive {
 				log.notify(`Removing clearRoom directive in ${this.pos.roomName}: operation completed.`);
 				this.remove();
 			}
+		// Clear path if controller is not reachable
+		} else if (this.room && this.room.creeps.length > 1) {
+			let currentlyDismantling = _.find(this.room.flags, function(flag) {
+				return (flag.color == DirectiveDismantle.color && flag.secondaryColor == DirectiveDismantle.secondaryColor)
```
Comment (bencbartlett) [inline src/directives/colony/clearRoom.ts]: use `DirectiveDismantle.find(this.room.flags)` to get a list of all dismantling directives in the room

File: src/directives/colony/clearRoom.ts [hunk 6]
```diff
@@ -83,6 +93,22 @@ export class DirectiveClearRoom extends Directive {
...
+			});
+
+			if (!currentlyDismantling) {
+				let pathablePos = this.room.creeps[0] ? this.room.creeps[0].pos
+					: Pathing.findPathablePosition(this.room.name);
+				let blockingLocation = Pathing.findBlockingPos(pathablePos, this.room.controller!.pos,
+					_.filter(this.room.structures, s => !s.isWalkable));
+				if (blockingLocation && blockingLocation.lookFor(LOOK_FLAGS).length <= 0) {
```
Comment (bencbartlett) [inline src/directives/colony/clearRoom.ts]: use Directive.isPresent(pos, 'pos')

File: src/utilities/utils.ts [hunk 7]
```diff
@@ -34,6 +34,15 @@ export function hasMinerals(store: { [resourceType: string]: number }): boolean
 	return false;
 }
 
+export function hasContents(store: { [resourceType: string]: number }): boolean {
```
Comment (bencbartlett) [inline src/utilities/utils.ts]: use storage/terminal.isEmpty

**Review by bencbartlett — CHANGES_REQUESTED** [review]

File: src/declarations/memory.d.ts (hunk 2 above)
Comment (MatthewARoy) [reply src/declarations/memory.d.ts:8]: I was thinking of setting some general principles for resource collection. Taking from all structures for example, and even mining walls.

As a more general logic, an increased resource collection could represent things like spawning haulers to pick up energy from SK room creep drops, collecting from invaders, general principles for mining, etc.
With how often overminds get wiped, starting from a blank slate you're desperate for any and all energy you can get a hold of. 
I had a grander vision for this but then realized it was out of scope, happy to remove from being a constant for now.

File: src/directives/colony/clearRoom.ts (hunk 4 above)
Comment (MatthewARoy) [reply src/directives/colony/clearRoom.ts]: Heh, it's definitely a useful one. I fixed some of them roughly in the power mining PR.

File: src/utilities/utils.ts (hunk 7 above)
Comment (MatthewARoy) [reply src/utilities/utils.ts]: Whoops...

**Pull request closed without merging**